ckanext/qgov/common/anti_csrf.py

```diff
@@ -278,38 +294,37 @@ def _get_post_token():
 
     This is normally a single 'token' parameter in the POST body.
     However, for compatibility with 'confirm-action' links,
-    it is also acceptable to provide the token as a query string parameter,
-    if there is no POST body.
+    it is also acceptable to provide the token as a query string parameter.
     """
-    if TOKEN_FIELD_NAME in request.environ['webob.adhoc_attrs']:
-        return request.token
-
-    # handle query string token if there are no POST parameters
-    # this is needed for the 'confirm-action' JavaScript module
-    if not request.POST and len(request.GET.getall(TOKEN_FIELD_NAME)) == 1:
-        request.token = request.GET.getone(TOKEN_FIELD_NAME)
-        del request.GET[TOKEN_FIELD_NAME]
-        return request.token
-
-    post_tokens = request.POST.getall(TOKEN_FIELD_NAME)
-    if not post_tokens:
-        csrf_fail("Missing CSRF token in form submission")
-    elif len(post_tokens) > 1:
-        csrf_fail("More than one CSRF token in form submission")
-    else:
-        request.token = post_tokens[0]
+    if TOKEN_FIELD_NAME in request_helpers.scoped_attrs():
+        return request_helpers.scoped_attrs()[TOKEN_FIELD_NAME]
 
-    # drop token from request so it doesn't populate resource extras
-    del request.POST[TOKEN_FIELD_NAME]
+    post_tokens = request_helpers.get_post_params(TOKEN_FIELD_NAME)
 
-    if not validate_token(request.token):
+    if post_tokens:
+        if len(post_tokens) > 1:
+            csrf_fail("More than one CSRF token in form submission")
+        else:
+            token = post_tokens[0]
+    else:
+        get_tokens = request_helpers.get_query_params(TOKEN_FIELD_NAME)
+        if len(get_tokens) == 1:
+            # handle query string token if there are no POST parameters
+            # this is needed for the 'confirm-action' JavaScript module
+            token = get_tokens[0]
+        else:
+            csrf_fail("Missing CSRF token in form submission")
+
+    if not validate_token(token):
         csrf_fail("Invalid token format")
 
-    return request.token
+    request_helpers.scoped_attrs()[TOKEN_FIELD_NAME] = token
+    request_helpers.delete_param(TOKEN_FIELD_NAME)
+    return token
 
 
 def intercept_csrf():
     """ Monkey-patch the core rendering methods to apply our CSRF tokens.
     """
-    base.render_jinja2 = anti_csrf_render_jinja2
+    base.render = anti_csrf_render
```

Comment:
not a fan of this. 
Is there no hooks already in the renderer core that we can't extend off (or see if they can add them {well us to our own version} )

Reply:
Nope, there are none. The entry point (which gets copied to the toolkit and handed out to plugins) is `base.render`, https://github.com/ckan/ckan/blob/master/ckan/lib/base.py#L120
which then calls native Pylons and Flask rendering functions and returns the result.